rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,770 @@
+contracted function we could nest them like so
+
+```rs
+fn my_div(dividend: u32, divisor: u32) -> u32 {
+  fn my_div_check_5e3713(dividend: u32, divisor: u32) -> u32 {
+    ...
+  }
+  ...
+}
+```
+
+This could be beneficial if we want to be able to allow contracts on trait impl
+items, in which case generating sibling functions is not allowed. On the other
+hand this makes it harder to implement contracts on traits *definitions*,
+because there is no body available which we could nest the function into.
+Utimately we may require both so that we vcan support both.
+
+
+What is required to make this work is an additional pass over the condition that
+replaces every `self` with a fresh identifier that now becomes the first
+argument of the function. In addition there are open questions as to how to
+resolve the nested name inside the compiler.
+
+### Explicit command line checking/substitution vs attributes: 
+
+Instead of
+  adding a new special `proof_for_contact` attributes we could have instead done:
+
+  1. **Check contracts on the command line** like CBMC does. This makes contract
+     checking a separate `kani` invocation with something like a
+     `--check-contract` flag that directs the system to instrument the function.
+     This is a very flexible design, but also easily used incorrectly.
+     Specifically nothing in the source indicates which harnesses are supposed
+     to be used for which contract, users must remember to invoke the check and
+     are also responsible for ensuring they really do verify *all* contacts they
+     will later be replacing and lastly.
+  2. **Check contracts with a `#[proof]` harness.** This would have used e.g. a
+     `#[for_contract]` attributes on a `#[proof]`. Since `#[for_contract]` is
+     *only* valid on a proof, we decided to just imply it and save the user some
+     headache. Contract checking harnesses are not meant to be reused for other
+     purposes anyway and if the user *really* wants to the can just factor out the
+     actual contents of the harness to reuse it.
+
+## Open questions
+
+<!-- For Developers -->
+
+- Is it really correct to return `kani::any()` from the replacement copy, even
+  if it can be a pointer?
+- Making result special. Should we use special syntax here like `@result` or
+  `kani::result()`, though with the latter I worry that people may get confused
+  because it is syntactic and not subject to usual `use` renaming and import
+  semantics. Alternatively we can let the user pick the name with an additional
+  argument to `ensures`, e.g. `ensures(my_result_var, CONDITION)`
+
+  Similar concers apply to `old`, which may be more appropriate to be special
+  syntax, e.g. `@old`.
+
+  See [#2597](https://github.com/model-checking/kani/issues/2597)
+- How to check the right contracts at the right time. By default `kani` and
+  `cargo kani` check all contracts in a file/workspace. This represents the
```

Comment:
This was deliberate to point to the case where you run `kani` on a single file. Should I still change it?

Reply:
Sorry, this is a misconception that I believe has spread in some of our documents. Standalone Kani runs on crates, not on single files, the same way that `rustc` compiles a crate not a single file. The file passed to Kani is the root file of the crate, and all the other files in the crate are found relative to the root file location.

For example, the test [kani/tests/kani/UnsizedCoercion/rc_outer_coercion.rs](https://github.com/model-checking/kani/blob/9f306999d26a965f912fd6e70533f7272fdc2b53/tests/kani/UnsizedCoercion/rc_outer_coercion.rs#L6) is a crate, which is verified with Kani. This crate has a module `defs` that points to file `defs.rs`.

Reply:
Yeah but I think that's confusing for the reader because you invoke standalone kani on a file. That that happens to then be interpreted as the root of the crate I find isn't an obvious leap. In my mind, when I think of crates I picture a directory.

I grant you that that conception is incorrect, but I'm saying it is the intuitive interpretation.